src/test/pcp_tests.cpp

```diff
@@ -0,0 +1,580 @@
+        }
+        return consume_bytes;
+    }
+
+    int Connect(const sockaddr* sa, socklen_t sa_len) const override {
+        CService service;
+        if (service.SetSockAddr(sa, sa_len) && service == CService(m_gateway_ip, 5351)) {
+            if (m_bound.IsBindAny()) { // If bind-any, bind to local ip.
+                m_bound = CService(m_local_ip, 0);
+            }
+            if (m_bound.GetPort() == 0) { // If no port assigned, assign port 1.
+                m_bound = CService(m_bound, 1);
+            }
+            m_connected = true;
+            return 0;
+        }
+        return -1;
+    }
+
+    int Bind(const sockaddr* sa, socklen_t sa_len) const override {
+        CService service;
+        if (service.SetSockAddr(sa, sa_len)) {
+            // Can only bind to one of our local ips
+            if (!service.IsBindAny() && service != m_local_ip) {
+                return -1;
+            }
+            m_bound = service;
+        }
+        return 0;
+    }
+
+    int Listen(int) const override { return -1; }
+
+    std::unique_ptr<Sock> Accept(sockaddr* addr, socklen_t* addr_len) const override
+    {
+        return nullptr;
+    };
+
+    int GetSockOpt(int level, int opt_name, void* opt_val, socklen_t* opt_len) const override
+    {
+        std::memset(opt_val, 0x0, *opt_len);
+        return 0;
+    }
+
+    int SetSockOpt(int, int, const void*, socklen_t) const override { return 0; }
+
+    int GetSockName(sockaddr* name, socklen_t* name_len) const override
+    {
+        // Return the address we've been bound to.
+        return m_bound.GetSockAddr(name, name_len) ? 0 : -1;
+    }
+
+    bool SetNonBlocking() const override { return true; }
+
+    bool IsSelectable() const override { return true; }
+
+    bool Wait(std::chrono::milliseconds timeout,
+              Event requested,
+              Event* occurred = nullptr) const override
+    {
+        // Only handles receive events.
+        if (AtEndOfScript() || requested != Sock::RECV) {
+            ElapseTime(timeout);
+        } else {
+            std::chrono::milliseconds delay = std::min(m_time_left, timeout);
+            ElapseTime(delay);
+            m_time_left -= delay;
+            if (CurOp().op == TestOp::RECV && m_time_left == 0s && occurred != nullptr) {
+                *occurred = Sock::RECV;
+            }
+            if (CurOp().op == TestOp::NOP) {
+                // This was a pure delay operation, move to the next op.
+                AdvanceOp();
+            }
+        }
+        return true;
+    }
+
+    bool WaitMany(std::chrono::milliseconds timeout, EventsPerSock& events_per_sock) const override
+    {
+        return false;
+    }
+
+    bool IsConnected(std::string&) const override
+    {
+        return true;
+    }
+
+private:
+    const std::vector<TestOp> m_script;
+    mutable size_t m_script_ptr = 0;
+    mutable std::chrono::milliseconds m_time_left;
+    mutable std::chrono::milliseconds m_time{1000};
```

Comment:
Correct. Thinking of it, we could define a constant for this in the MockableClock like `INITIAL_MOCK_TIME`. This would document it at the same time as preventing that mistake being made in future tests that use the mockable steady clock.

The alternative was to make the mock time in MockableClock an optional, but this would mean it could no longer simply be an atomic value, so was kind of a hassle compared to starting at non-zero.